File: app/models/analytics.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class PostAnalytics(Base):
    __tablename__ = "post_analytics"

    id: Mapped[int] = mapped_column(primary_key=True)
    post_id: Mapped[int] = mapped_column(ForeignKey("posts.id", ondelete="CASCADE"))
    platform: Mapped[str] = mapped_column(String(32))
    fetched_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    likes: Mapped[int] = mapped_column(Integer, default=0)
    reposts: Mapped[int] = mapped_column(Integer, default=0)
    replies: Mapped[int] = mapped_column(Integer, default=0)
    views: Mapped[int] = mapped_column(Integer, default=0)
    clicks: Mapped[int] = mapped_column(Integer, default=0)
    raw_data: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON

    post: Mapped["Post"] = relationship(back_populates="analytics")  # noqa: F821


class AnalyticsReport(Base):
    __tablename__ = "analytics_reports"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"))
    campaign_id: Mapped[int | None] = mapped_column(
        ForeignKey("campaigns.id", ondelete="SET NULL"), nullable=True
    )
    title: Mapped[str] = mapped_column(String(255))
    report_text: Mapped[str] = mapped_column(Text)  # Claude-generated markdown
    date_range_start: Mapped[datetime] = mapped_column(DateTime)
    date_range_end: Mapped[datetime] = mapped_column(DateTime)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    campaign: Mapped["Campaign | None"] = relationship(  # noqa: F821
        back_populates="analytics_reports"
    )

File: client/src/api/analytics.ts
import { apiFetch } from "./client";

export interface AnalyticsEntry {
  id: number;
  postId: number;
  userId: number;
  platform: string;
  fetchedAt: string;
  likes: number;
  shares: number;
  comments: number;
  views: number;
  clicks: number;
  rawData: Record<string, unknown> | null;
}

export const analytics = {
  list: () => apiFetch<AnalyticsEntry[]>("/api/analytics"),
  fetch: (postId: number) =>
    apiFetch<AnalyticsEntry>(`/api/analytics/fetch/${postId}`, { method: "POST" }),
  report: (campaignId?: number, campaignName?: string) =>
    apiFetch<{ report: string }>("/api/analytics/report", {
      method: "POST",
      body: JSON.stringify({ campaignId, campaignName }),
    }),
};
